src/wp-includes/class-wp-block-supports.php

```diff
@@ -0,0 +1,196 @@
+				foreach ( $new_attributes as $attribute_name => $attribute_value ) {
+					if ( empty( $output[ $attribute_name ] ) ) {
+						$output[ $attribute_name ] = $attribute_value;
+					} else {
+						$output[ $attribute_name ] .= " $attribute_value";
+					}
+				}
+			}
+		}
+
+		return $output;
+	}
+
+	/**
+	 * Registers the block attributes required by the different block supports.
+	 */
+	private function register_attributes() {
+		$block_registry         = WP_Block_Type_Registry::get_instance();
+		$registered_block_types = $block_registry->get_all_registered();
+		foreach ( $registered_block_types as $block_type ) {
+			if ( ! property_exists( $block_type, 'supports' ) ) {
+				continue;
+			}
+			if ( ! $block_type->attributes ) {
+				$block_type->attributes = array();
+			}
+
+			foreach ( $this->block_supports as $name => $block_support_config ) {
+				if ( ! isset( $block_support_config['register_attribute'] ) ) {
+					continue;
+				}
+
+				call_user_func(
+					$block_support_config['register_attribute'],
+					$block_type
+				);
+			}
+		}
+	}
+}
+
+/**
+ * Generates a string of attributes by applying to the current block being
+ * rendered all of the features that the block supports.
+ *
+ * @param array $extra_attributes Optional. Extra attributes to render on the block wrapper.
+ *
+ * @return string String of HTML classes.
+ */
+function get_block_wrapper_attributes( $extra_attributes = array() ) {
+	global $current_parsed_block;
+	$new_attributes = WP_Block_Supports::get_instance()->apply_block_supports( $current_parsed_block );
+
+	if ( empty( $new_attributes ) && empty( $extra_attributes ) ) {
+		return '';
+	}
+
+	// This is hardcoded on purpose.
+	// We only support a fixed list of attributes.
+	$attributes_to_merge = array( 'style', 'class' );
+	$attributes          = array();
+	foreach ( $attributes_to_merge as $attribute_name ) {
+		if ( empty( $new_attributes[ $attribute_name ] ) && empty( $extra_attributes[ $attribute_name ] ) ) {
+			continue;
+		}
+
+		if ( empty( $new_attributes[ $attribute_name ] ) ) {
+			$attributes[ $attribute_name ] = $extra_attributes[ $attribute_name ];
+			continue;
+		}
+
+		if ( empty( $extra_attributes[ $attribute_name ] ) ) {
+			$attributes[ $attribute_name ] = $new_attributes[ $attribute_name ];
+			continue;
+		}
+
+		$attributes[ $attribute_name ] = $extra_attributes[ $attribute_name ] . ' ' . $new_attributes[ $attribute_name ];
+	}
+
+	foreach ( $extra_attributes as $attribute_name => $value ) {
+		if ( ! in_array( $attribute_name, $attributes_to_merge, true ) ) {
+			$attributes[ $attribute_name ] = $value;
+		}
+	}
+
+	if ( empty( $attributes ) ) {
+		return '';
+	}
+
+	$normalized_attributes = array();
+	foreach ( $attributes as $key => $value ) {
+		$normalized_attributes[] = $key . '="' . esc_attr( $value ) . '"';
+	}
+
+	return implode( ' ', $normalized_attributes );
+}
+
+add_action( 'init', array( 'WP_Block_Supports', 'init' ), 22 );
```

Comment:
PHP files beginning with the `class-` prefix should define the class and nothing else. Let's move this `add_action()` somewhere else. Maybe `src/wp-includes/default-filters.php`.